### ISSUE TITLE
Why is .register returning cached certificates?

### ISSUE THREAD

**Issue by manast** [Daplie/node-letsencrypt#62]
I have recently discovered that .register also returns an existing certificate if not expired. This seems to be inconsistent with the rest API, for instance the .check method, and the example in the README, where it seems that register will force a registration (note the comment   ```// Register Certificate manually```). Furthermore, I guess that the most common pattern is to renew certificates several days before it really gets expired, what would be the correct API to call in order to force a certificate renewal?

```
// Check in-memory cache of certificates for the named domain
le.check({ domains: [ 'example.com' ] }).then(function (results) {
  if (results) {
    // we already have certificates
    return;
  }


  // Register Certificate manually
  le.register({

    domains: ['example.com']                                // CHANGE TO YOUR DOMAIN (list for SANS)
  , email: 'user@email.com'                                 // CHANGE TO YOUR EMAIL
  , agreeTos: ''                                            // set to tosUrl string (or true) to pre-approve (and skip agreeToTerms)
  , rsaKeySize: 2048                                        // 2048 or higher
  , challengeType: 'http-01'                                // http-01, tls-sni-01, or dns-01

  }).then(function (results) {

    console.log('success');

  }, function (err) {

    // Note: you must either use le.middleware() with express,
    // manually use le.challenges['http-01'].get(opts, domain, key, val, done)
    // or have a webserver running and responding
    // to /.well-known/acme-challenge at `webrootPath`
    console.error('[Error]: node-letsencrypt/examples/standalone');
    console.error(err.stack);

  });

});
```

**Comment by insightfuls** [comment]
It will renew a few days early by default.

But you can pass `renewBy` and/or `renewWithin` to adjust this. These are documented in the https://github.com/Daplie/letsencrypt-express README, and unfortunately not here.

You can also pass `duplicate:true` to force duplicate registration/early renewal, which doesn't seem to be documented anywhere, except in a roundabout way in https://github.com/Daplie/letsencrypt-cli/ with some discussion of it at https://github.com/Daplie/letsencrypt-cli/issues/22

Pull requests to improve the documentation are welcome.

**Comment by manast** [comment]
@insightfuls I am still a bit confused about how the api is supposed to work. 

According to the example above (from the README), first a call to check is done, if we get a valid certificate then returns, otherwise it calls to ```.register```. Now, in my code, I want to call ```register``` some time before the certificate expiration, but instead of renewing the certificate it is returning the old one. So I get that I should also send the option ```renewBy```/```renewWithin```, but then, could I just skip the call to ```check```above,  or is ```register``` still doing an extra call to the ACME server and getting the old certificate whereas ```check``` is more efficient by just returning a local cached one?

**Comment by insightfuls** [comment]
Yeah, the semantics of `check` and `register` are a little mixed up. However, here's what happens.

`check` will just check the cache and return a certificate or not.

`register` will:
- use the same function as `check` uses to identify if there is a cached cert already
- register and return a new cert if there is none
- return a cached cert if there is one
- if the cached cert is 'renewable' (due to expire within `renewWithin` or `duplicate` was specified), initiate a renewal in the background, and return a promise for the renewal in `_renewing` (along with the cached cert); this is undocumented but fairly important, and we probably need to clean it up somehow at some stage

Consequently:
- calling `register` prior to the certificate becoming renewable will not initiate renewal; it will do the same as `check`, which is somewhat unexpected
- even if you force renewal with `duplicate`, `register` will still return the cached cert, which is fairly unexpected; if the response contains `_renewing`, watch it if you want the new cert
- calling `register` multiple times after a certificate becomes renewable will initiate multiple renewals in the background; rate limits can be hit by doing this; renewals that fail will probably become unhandled rejections

There is another function `renew` that probably does more what you want:

- if the cert is not renewable, fail
- if the cert is renewable, return a new one

It accepts `duplicate` and `renewWithin` just like `register` does to determine if a cert is renewable.

```
le.check({ domains: domains }).then(function(cert) {
  if (cert) {
    // we have a certificate; renew it
    return le.renew({ duplicate: true }, cert);
  }

  // register a certificate manually
  le.register({
    domains: domains
  , email: 'user@email.com'
  , agreeTos: ''
  , rsaKeySize: 2048
  , challengeType: 'http-01'
  }).then(function(cert) {
    console.log('got cert for ' + cert.domains.join(", "));
  }, function(err) {
    console.log('failed: ' + err);
  });
});
```

**Comment by insightfuls** [comment]
O, and I was very sloppy above. Whenever I've said "return" I mean fulfill the returned promise; likewise "fail" means reject the returned promise.

**Comment by manast** [comment]
@insightfuls that was a very useful answer, thank you very much. Specifically the ```renew```method is the silver bullet for my case.

**Comment by coolaj86** [comment]
FYI the official repo is at https://git.coolaj86.com/coolaj86/greenlock.js

Latest version supports Let's Encrypt v2 :)
